pkg/endpoint/endpoint.go

```diff
@@ -2403,3 +2433,61 @@ func (e *Endpoint) IPs() []net.IP {
 func (e *Endpoint) InsertEvent() {
 	e.getLogger().Info("New endpoint")
 }
+
+// must be called with endpoint mutex locked.
+func (e *Endpoint) syncPolicyMap() error {
+
+	if e.realizedMapState == nil {
+		e.realizedMapState = make(map[policymap.PolicyKey]struct{})
+	}
+
+	currentMapContents, err := e.PolicyMap.DumpToSlice()
+
+	if err != nil {
+		return fmt.Errorf("unable to dump PolicyMap for endpoint: %s", err)
+	}
+
+	for _, entry := range currentMapContents {
+		keyToAnalyze := policymap.PolicyKey{
+			Identity:         entry.Key.GetIdentity(),
+			DestPort:         byteorder.NetworkToHost(entry.Key.GetPort()).(uint16),
+			Nexthdr:          entry.Key.GetProto(),
+			TrafficDirection: entry.Key.GetDirection(),
+		}
+
+		// If key that is in policy map is not in desired state, just remove it.
+		if _, ok := e.desiredMapState[keyToAnalyze]; !ok {
+			err := e.PolicyMap.Delete(keyToAnalyze.Identity, keyToAnalyze.DestPort, u8proto.U8proto(keyToAnalyze.Nexthdr), policymap.TrafficDirection(keyToAnalyze.TrafficDirection))
+			if err != nil {
+				return err
```

Comment:
I suppose that this is probably how we're already handling errors, but I wonder if it's better to attempt to put as much of the desired state in place as possible rather than erroring out at the first failure? Similarly for the `Allow()` call below.

I agree that #3897 should fix this once and for all, but maybe the failure conditions here can be made a bit safer with a minor incremental change?

Reply:
I'll look into doing this as a follow-up commit in this PR.